### ISSUE TITLE
Need ability to redact data at the check level

### ISSUE THREAD

**Issue by CVTJNII** [sensu/sensu#1513]
I'm using Chef to manage Sensu so Chef takes care of all my attribute handling.  I currently have several checks that accept passwords on the command line which Uchiwa merrily displays to the world.  In looking over the docs I do not see any way to redact passwords on a per check basis.  Sensu supports 'redact', but that is at the client level and I don't want to redact all commands, only commands with passwords.  There are tokens, where I could name a token password and set the password that way, but per https://sensuapp.org/docs/0.24/reference/checks.html#check-token-substitution tokens are also per client.  I may be able to globally blacklist 'password' and declare all passwords as :::password|#{actual_password}::: but this is clunky and opens the risk of bad replacement if password is ever set at the client level at the future (Assuming it works, I've pinned to Uchiwa 0.14.5 due to silence stash changes which still displays :::password|#{actual_password}::: with redact: [ "password" ] set in the client config.)

What I believe would be best is to set redact at the check level, so I can, in effect, say "don't display the command for this check, it contains passwords".  This would be especially useful important for "remote check" hosts that are responsible for acting as a checker pool to remotely check many hosts.

## Expected Behavior
Ability to set attribute redaction at a per-check level

## Current Behavior
Attributes can only be redacted at a per client level.

## Possible Solution
Allow redact to be set in the check scope.

**Comment by CVTJNII** [comment]
@cwjohnston and I had a chat about this in #sensu on IRC where we discussed the spirit of the request:

> <cwj> CVTJNII: FWIW I am working on a response to your issue re: token substitution and redaction at check level. I may be mistaken but I suspect that you misunderstand how the existing token substitution and redaction work. I don't know if portertech has looked at the issue, but I think its unlikely that adding that feature at the check level is going to be
> <cwj> accepted upstream.
> <CVTJNII> Hrm, okay, then how should I redact such things in Uchiwa?
> <CVTJNII> There is an old issue for this problem that was closed with "use redaction"
> <CVTJNII> And I very well may misunderstand it, the issue was found during a deployment so that issue was created under duress.
> <cwj> CVTJNII: stand by, i'm creating a gist which I hope will be illustrative
> <CVTJNII> Okay, thanks
> <cwj> CVTJNII: ok try this on https://gist.github.com/cwjohnston/5d9bb797d854e819d11d38c6d295165f
> <CVTJNII> How does that work with checks scheduled by the server?
> <cwj> the same way
> <CVTJNII> So all my remote checkers need to know the settings for the checks they're going to run?
> <cwj> yes
> <CVTJNII> That's a mess
> <CVTJNII> Sorry, got to be honest
> <cwj> thats ok, we can disagree
> <CVTJNII> That's really the spirit of my request.  I think this kind of data, as it's data for the check, should be in check scope, not client scope.
> <CVTJNII> Using this method will make my config orders of magnitude more complex as I'll have to create unique variable names per check and distribute them to all hosts which may run that check.
> <CVTJNII> If I could say "Hey, don't display this attribute for this one check" that would be very nice
> <cwj> but if you have a check which is the same across many environments, except for the password, you now need a different check definition for each environment
> <CVTJNII> But Chef manages that all for me
> <cwj> right, and from my perspective Chef can also manage client configuration tokens
> <CVTJNII> I guess my thought is using tokens is very complicated when all I want to do is not display the command for checks that have passwords in the command.
> <cwj> that sounds like it would make it harder to diagnose configuration mismatches and other problems if we can't see the basic command thats being run by the client
> <cwj> whereas the token substitution model at least lets us see the command, flags, etc minus the actual token values
> <cwj> you are certainly welcome to submit a PR, I'm just trying to clear up any potential misunderstandings before you spend time on something that might not be accepted upstream
> <CVTJNII> That is true, so I mean some sort of a password/secret flag in the command itself would be even better
> <CVTJNII> I chose redaction at the [check] level as I thought it would be easier, because it's already there and understood by Uchiwa
> <CVTJNII> That's where I was going with :::password|actual_password:::, basically treating the password token as a "don't display this" flag
> <CVTJNII> But that didn't work in my very short test, though I didn't dive in to figure out why as I didn't have the time.
> <CVTJNII> So maybe that's a better direction, if per check redaction isn't as easy as I assumed it to be.
> <CVTJNII> cwj: In any case thanks very much for the advice, it is very much appreciated.
> <cwj> CVTJNII: no problem, hope you are able to find a workable solution either way! 😉
> <CVTJNII> Mind if I copy this into the issue as it's very relevant?
> <cwj> feel free :)

tl;dr: In my config I'd like a way to redact passwords at the check config level, as tokens are a force fit for me due to how I'm managing the config.  I picked redact at the check scope as an easy solution, but maybe a "obfuscate this" flag would be better and meet the spirit of the request?

**Comment by portertech** [comment]
Sensu client definition attribute redaction will stop sensitive information from leaving the local machine, it will not be logged or sent over the Sensu Transport (RabbitMQ). Check definition attribute redaction doesn't make much sense, as the original values (non-redacted) need to be sent over the transport in order to be used when executing the check.

**Comment by CVTJNII** [comment]
I'm not concerned about the transport as the transport is encrypted, performs SSL certificate verification, and requires a password.  I am concerned about the Uchiwa UI which is open to users to view status and suppress alerts.  I'm looking for a solution that allows me to mark a field as sensitive so that it can be displayed to the outside world without needing a complex set of configuration dependencies.

As I was discussing with CWJohnson I'm strongly opposed to defining these settings at the client scope in my environment.  We have a large number of checks configured by multiple teams which are executed by pools of workers.  To implement the solution you're proposing every pool worker needs to be configured with the settings of every check they could need to run.  This increases the chance of error by at least an order of magnitude, as it obfuscates the configuration, requires the configuration to be in sync for at least double the number of servers, and requires Chef (and users) to generate a much more complicated configuration.  Right now my configuration only needs to be correct in one place on the Sensu Servers, however with this method the configuration needs to be correct on all the Sensu servers and all the Sensu Clients which seems like a recipe for false errors and reduced reliability due to increased complexity to me.  As the other benefits of token substitution aren't needed in our otherwise simple config, this seems like a bit of an anti-pattern for this problem.

I consider my queues, and as such the roundrobin queues which drive check execution, to be secure.  While my results queue is also secure the output of it is not, and that's what I'm trying to protect without the risk of a very complex configuration.

**Comment by CVTJNII** [comment]
Playing with the official solution a bit more I see some potential issues with how it redacts data.  Uchiwa will display the client config to the user, including any token definitions, in it's entirety unless the key is named "password".  For example:

![screen shot 2017-03-06 at 5 39 27 pm](https://cloud.githubusercontent.com/assets/13871930/23633907/a04c03d0-0295-11e7-859a-2fd0d7715075.png)

So if I store my secret in a path that contains "password", then it will be hidden, but otherwise it will still be viewable.  So while functional, the recommended solution seems to have been designed not with passwords in mind.  Because of this I'm still of the opinion it would be nice to see some alternatives configuration options more tailored to secrets and passwords.

As a postscript for others with similar concerns, while troubleshooting I found that the following config is completely legal:
```json
"some-check": {
  "command": "check-command $(cat /tmp/testfile)"
}
```

This is probably the direction I'm going to move in my enterprise config as it removes the password from Sensu's care completely, thereby bypassing all these issues, and opens up using a password management wrapper to store the passwords centrally and securely.

**Comment by portertech** [comment]
@CVTJNII the client attribute `redact` can be used to change the attributes that have their values redacted, https://sensuapp.org/docs/0.28/reference/clients.html#client-attributes

**Comment by majormoses** [comment]
The redact condif at client level solution seems right to me, in general we should attempt to remove all sensitive cli options passed as even if we do this in sensu it can still be seen by any user on the system by simply running `ps`. Good examples of this are say the chef client checks (rather than giving a key we give it a path to a key which should be locked down to sudoers and sensu) or the aws checks that use IAM correctly.

**Comment by mfornasa** [comment]
In my setup I check remote http sites using proxy clients and roundrobin executors. So, I don't have any client definition.
I need to pass secret information to the checks. What is the standard solution?

**Comment by majormoses** [comment]
I will be honest I do not use proxy hosts myself so I am not sure the best way to do that. Without an enhancement I have two possible suggestions:
- drop a file on the host that it will eventually land on and in the check read from the file. (much like what was suggested above, the chef checks, and now most of the aws checks).
- configure an ENV var for the sensu client process and read that ENV var in the check. This could be a bit tricker to do SAFELY.

**Comment by majormoses** [comment]
> I'm not concerned about the transport as the transport is encrypted, performs SSL certificate verification, and requires a password.

You can use redis as a transport and despite my warnings people run it in production.

**Comment by majormoses** [comment]
> As a postscript for others with similar concerns, while troubleshooting I found that the following config is completely legal:
"some-check": {
  "command": "check-command $(cat /tmp/testfile)"
}
This is probably the direction I'm going to move in my enterprise config as it removes the password from Sensu's care completely, thereby bypassing all these issues, and opens up using a password management wrapper to store the passwords centrally and securely.

This is definitely one of the best options IMHO, token substitution works great for it's intended use cases and supported (even if misguided) architectures.

**Comment by portertech** [comment]
For cases not covered by token substitution and client redaction, out-of-band methods are most effective, e.g. $ENV_VAR or $(cat /tmp/testfile).

**Comment by majormoses** [comment]
@portertech I have been steering people away from using $ENV_VAR as they tend to miss the point and do this `MY_SECRET=foo check-command $MY_SECRET` which is not any better.
